src/phoenix/core/traces.py

```diff
@@ -1,108 +1,288 @@
-CUMULATIVE_LLM_TOKEN_COUNT_TOTAL = "__computed__cumulative_token_count_total"
-CUMULATIVE_LLM_TOKEN_COUNT_PROMPT = "__computed__cumulative_token_count_prompt"
-CUMULATIVE_LLM_TOKEN_COUNT_COMPLETION = "__computed__cumulative_token_count_completion"
+COMPUTED_NUMERIC_PREFIX = "__computed_numeric__."
+LATENCY_MS = COMPUTED_NUMERIC_PREFIX + "latency_ms"
+"The latency (or duration) of the span in milliseconds"
+CUMULATIVE_LLM_TOKEN_COUNT_TOTAL = COMPUTED_NUMERIC_PREFIX + "cumulative_token_count_total"
+CUMULATIVE_LLM_TOKEN_COUNT_PROMPT = COMPUTED_NUMERIC_PREFIX + "cumulative_token_count_prompt"
+CUMULATIVE_LLM_TOKEN_COUNT_COMPLETION = (
+    COMPUTED_NUMERIC_PREFIX + "cumulative_token_count_completion"
+)
+
+
+class ReadableSpan(ObjectProxy):  # type: ignore
+    __wrapped__: Span
+
+    def __init__(self, span: Span) -> None:
+        super().__init__(span)
+        self._self_computed_numeric_values: Dict[str, SupportsFloat] = {}
+
+    def __getitem__(self, key: str) -> Any:
+        if key.startswith(COMPUTED_NUMERIC_PREFIX):
+            return self._self_computed_numeric_values.get(key)
+        if key.startswith(CONTEXT_PREFIX):
+            return getattr(
+                self.__wrapped__.context,
+                key[len(CONTEXT_PREFIX) :],
+                None,
+            )
+        if key.startswith(ATTRIBUTE_PREFIX):
+            return self.__wrapped__.attributes.get(
+                key[len(ATTRIBUTE_PREFIX) :],
+            )
+        return getattr(self.__wrapped__, key, None)
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        if not key.startswith(COMPUTED_NUMERIC_PREFIX):
+            raise KeyError(f"{key} is not a computed value")
+        self._self_computed_numeric_values[key] = value
+
+
+ParentSpanID: TypeAlias = SpanID
+ChildSpanID: TypeAlias = SpanID
 
 
 class Traces:
-    """
-    Traces class is used to contain abstractions around the traces dataframe
-    """
-
-    def __init__(self, df: pd.DataFrame):
-        df = df.set_index(SPAN_ID, drop=False)
-        df[LATENCY_MS] = (df[END_TIME] - df[START_TIME]).dt.total_seconds() * 1000
-        parent_span_ids = df.loc[:, PARENT_ID]
-        for column, cumulative_column in (
+    def __init__(
+        self,
+        spans: Optional[Iterable[Span]] = None,
+    ) -> None:
+        self._queue: "SimpleQueue[Optional[Span]]" = SimpleQueue()
+        # Putting `None` as the sentinel value for queue termination.
+        weakref.finalize(self, self._queue.put, None)
+        for span in spans or ():
+            self._queue.put(span)
+        self._lock = Lock()
+        self._spans: Dict[SpanID, ReadableSpan] = {}
+        self._parent_span_ids: Dict[SpanID, ParentSpanID] = {}
+        self._traces: Dict[TraceID, List[SpanID]] = defaultdict(list)
+        self._child_span_ids: DefaultDict[SpanID, List[ChildSpanID]] = defaultdict(list)
+        self._orphan_spans: DefaultDict[ParentSpanID, List[Span]] = defaultdict(list)
```

Comment:
Nit: These names sound more like lists than dictionaries.

Reply:
Dicts and lists are very similar in that lists are just dicts with integer keys.